src/search/domain_abstractions/split.cc

```diff
@@ -0,0 +1,117 @@
+            getEnumForString(method)), log(log) {
+        log << "Set SPLIT METHOD: " << method << endl;
+    }
+
+    VariableGroupVectors
+    DomainSplitter::split(const shared_ptr <Flaw> &flaw, const shared_ptr <DomainAbstraction> &currentAbstraction) {
+        // Select method on how to split and call Submethod. We return the new Abstraction
+        if (currentMethod == SplitMethod::HARDSPLIT) {
+            return performHardSplit(flaw, currentAbstraction);
+        } else if (currentMethod == SplitMethod::EVENSPLIT) {
+            return performEvenSplit(flaw, currentAbstraction);
+        } else {
+            // default fallback if unknown
+            //log << "CEGAR -- Split: using fallback Method(Hard-Split) since given method not specified!" << endl;
+            return performHardSplit(flaw, currentAbstraction);
+        }
+    }
+
+    VariableGroupVectors
+    DomainSplitter::performHardSplit(const shared_ptr <Flaw> &flaw,
+                                     const shared_ptr <DomainAbstraction> &currentAbstraction) {
+        /*
+         * Split missed fact from rest of facts in same group for every variable (that has missed facts in their domain)
+         * */
+        VariableGroupVectors oldAbstraction = currentAbstraction->getAbstractDomains();
+        VariableGroupVectors newAbstraction = oldAbstraction;
+
+        vector<int> stateValues = flaw->getStateWhereFlawHappensCopy();
+        vector<int> abstractStateWhereFlawHappened = currentAbstraction->getGroupAssignmentsForConcreteState(
+                stateValues);
+        // TODO: we assume that there is max one missed fact for every variable!? -> YES in the formalism used in downward yes!
+        shared_ptr<vector<FactPair>> missedFacts = flaw->missedFacts;
+        assert(!missedFacts->empty());
+        // loop over variables and their domains
+        for (auto factPair: *missedFacts) {
+            VariableGroupVector oldVariableAbstractDomain = oldAbstraction.at(factPair.var);
+            VariableGroupVector newVariableAbstractDomain = oldVariableAbstractDomain;
+            int maxGroupNumber = *max_element(oldVariableAbstractDomain.begin(), oldVariableAbstractDomain.end());
+            assert(maxGroupNumber <= (int) oldVariableAbstractDomain.size());
+            // just give missed-fact a new group TODO: if check is already in separate group flaw should not have been happened!
+            //int domainIndex = currentAbstraction->getDomainIndexOfVariableValue(factPair.var, factPair.value);
+            //assert(domainIndex == factPair.value);
+            newVariableAbstractDomain[factPair.value] = maxGroupNumber + 1;
+            newAbstraction[factPair.var] = newVariableAbstractDomain;
+        }
+        assert(newAbstraction.size() == oldAbstraction.size());
+        return newAbstraction;
+    }
+
+    VariableGroupVectors DomainSplitter::performEvenSplit(const shared_ptr <Flaw> &flaw,
```

Comment:
Maybe rename to `performRandomUniformSplit`?

Reply:
Great Idea, done!